### PR TITLE
Geometry_Engine: Add methods to assist with tagging workflows

### LINKED ISSUE #3033: Additional methods for Geometry_Engine
#### Description:

We need a few additional methods to handle new use cases such as:
- Get the smallest possible angle between 2 vectors
- Thickening a line into a polyline
- Find line intersections between a polyline and a line
- Find distance between 2 points along a direction vector

### LINKED ISSUE #3034: Additional methods for Bhom_Engine
#### Description:

We need a few additional methods to handle new use cases such as:
- Converting between degree and radian values
- Copy property values between 2 objects where the property names match
- Shifting and splitting a list

### PR DESCRIPTION
**Pull request by vietle-bh**
### Issues addressed by this PR

Closes #3033, Closes #3034

Added new methods such as:

-   Get the smallest possible angle between 2 vectors
-   Thickening a line into a polyline
-   Find line intersections between a polyline and a line
-   Find the distance between 2 points along a direction vector
-   Converting between degree and radian values
-   Copy property values between 2 objects where the property names match
-   Shifting and splitting a list

### Test files
[Test_script.zip](https://github.com/BHoM/BHoM_Engine/files/12428371/Test_script.zip)

### Changelog

Added engine methods for new use cases.

### CODE REVIEW

**Comment by vietle-bh** [comment]
@BHoMBot check compliance
@BHoMBot check core
@BHoMBot check null-handling
@BHoMBot check serialisation

**Comment by vietle-bh** [comment]
@BHoMBot check unit-tests
@BHoMBot check versioning

**Comment by vietle-bh** [comment]
@BHoMBot check unit-tests

File: BHoM_Engine/Modify/SplitAndRemoveAtIndexes.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    {
+        /***************************************************/
+        /****               Public Methods              ****/
+        /***************************************************/
+
+        [Description("Removes list items at given indexes, then returns the remaining objects as sublists of consecutive items.")]
+        [Input("items", "A list of items to split at one or more indexes.")]
+        [Input("indexes", "Indexes of items to remove.")]
+        [Output("lists", "Sublists of consecutive items that remain after items at input indexes have been removed.")]
+        public static List<List<T>> SplitAndRemoveAtIndexes<T>(this List<T> items, List<int> indexes)
```
Comment (pawelbaran) [inline BHoM_Engine/Modify/SplitAndRemoveAtIndexes.cs:40]: I would rather name the method as `RemoveAndSplitAtIndexes`

File: Geometry_Engine/Convert/ToXY.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+using BH.oM.Base.Attributes;
+using BH.oM.Geometry;
+using System.ComponentModel;
+
+namespace BH.Engine.Geometry
+{
+    public static partial class Convert
```
Comment (pawelbaran) [inline Geometry_Engine/Convert/ToXY.cs:29]: This is definitely not a `Convert` method - a suitable method for this class takes an object of type A and converts it to an equivalent represented by type B. Instead, here we do a projection of `Point`, which definitely is `Modify`.

File: Geometry_Engine/Convert/ToXY.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    {
+        /***************************************************/
+        /****               Public Methods              ****/
+        /***************************************************/
+
+        [Description("Returns the projection of a point on the XY plane. This shorthand method should run quicker than BH.Engine.Geometry.Project(this Point, Plane) in performance-sensitive applications.")]
+        [Input("pnt", "A point to project onto the XY plane.")]
+        [Output("pntOnXY", "Projection of the input point on the XY plane.")]
+        public static Point ToXY(this Point pnt)
```
Comment (pawelbaran) [inline Geometry_Engine/Convert/ToXY.cs:38]: I would rather call this method `ProjectToXY`

**Review by pawelbaran — CHANGES_REQUESTED** [review]
I like the code, only a few minor tweaks suggested. However, please be so kind and provide a Grasshopper test script with the added methods - you are touching core projects of the entire BHoM here so extensive testing is a must. The script should ensure correctness of the methods as it will be serialised to serve as a unit test in the future.

**Comment by vietle-bh** [comment]
@BHoMBot check compliance
@BHoMBot check core
@BHoMBot check null-handling
@BHoMBot check serialisation

**Comment by vietle-bh** [comment]
@BHoMBot check compliance
@BHoMBot check core
@BHoMBot check null-handling
@BHoMBot check serialisation

**Comment by vietle-bh** [comment]
@BHoMBot check unit-tests

**Comment by vietle-bh** [comment]
@pawelbaran , I've added the GH test script to this PR's description.

**Comment by pawelbaran** [comment]
@vietle-bh I have had a look at the test script you prepared - it is a good start, but definitely not sufficient (not enough cases, no verification). I quickly patched up an alternative how I would do it and uploaded it [here](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?RootFolder=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F02%5FPull%20Request%2FBHoM%2FBHoM%5FEngine%2FGeometry%5FEngine%2F%233102%2DAddTaggingTool&FolderCTID=0x0120008122C8891F89054B8ACED0196C70DFC4). Let's have a chat tomorrow to coordinate on how to bring this PR over the line 💪

**Comment by pawelbaran** [comment]
@BHoMBot check unit-tests

**Review by pawelbaran — APPROVED** [review]
Happy to approve the PR following the code review, testing and unit test creation done together with @vietle-bh 👍 Good stuff!

**Comment by vietle-bh** [comment]
@BHoMBot check required

**Comment by vietle-bh** [comment]
@BHoMBot check versioning

**Comment by vietle-bh** [comment]
Sorry @FraserGreenroyd , similar story here where the failed test isn't relevant to this PR. It's the unit-tests this time. Could we bypass this please?

**Comment by FraserGreenroyd** [comment]
@BHoMBot check compliance

**Comment by FraserGreenroyd** [comment]
@BHoMBot check core
@BHoMBot check null-handling
@BHoMBot check serialisation

**Comment by FraserGreenroyd** [comment]
@BHoMBot check versioning
@BHoMBot check installer

**Comment by FraserGreenroyd** [comment]
@BHoMBot this is a DevOps instruction. I am requesting neutral checks on: ready-to-merge, unit-tests

**Pull request merged**